nodejs/aws-infra/experimental/clusterAutoScaling.ts

```diff
@@ -0,0 +1,566 @@
+                echo -e \"$DIR_SRC:/ $DIR_TGT nfs4 nfsvers=4.1,rsize=1048576,wsize=1048576,hard,timeo=600,retrans=2 0 0\" | tee -a /etc/fstab
+                mount -a -t nfs4
+                # Restart Docker
+                docker ps
+                service docker stop
+                service docker start
+            `;
+        }
+
+        return `#cloud-config
+        repo_upgrade_exclude:
+            - kernel*
+        packages:
+            - aws-cfn-bootstrap
+            - aws-cli
+            - nfs-utils
+        mounts:
+            - ['/dev/xvdb', 'none', 'swap', 'sw', '0', '0']
+        bootcmd:
+            - mkswap /dev/xvdb
+            - swapon /dev/xvdb
+            - echo ECS_CLUSTER='${clusterId}' >> /etc/ecs/ecs.config
+            - echo ECS_ENGINE_AUTH_TYPE=docker >> /etc/ecs/ecs.config
+        runcmd:
+            # Set and use variables in the same command, since it's not obvious if
+            # different commands will run in different shells.
+            - |
+                # Knock one letter off of availability zone to get region.
+                AWS_AVAILABILITY_ZONE=$(curl -s 169.254.169.254/2016-09-02/meta-data/placement/availability-zone)
+                AWS_REGION=$(echo $AWS_AVAILABILITY_ZONE | sed 's/.$//')
+
+                ${fileSystemRuncmdBlock}
+
+                # Disable container access to EC2 metadata instance
+                # See http://docs.aws.amazon.com/AmazonECS/latest/developerguide/instance_IAM_role.html
+                iptables --insert FORWARD 1 --in-interface docker+ --destination 169.254.169.254/32 --jump DROP
+                service iptables save
+
+                /opt/aws/bin/cfn-signal \
+                    --region "\${AWS_REGION}" \
+                    --stack "${cloudFormationStackName}" \
+                    --resource Instances
+        `;
+    });
+}
+
+export class ClusterAutoScalingGroup extends pulumi.ComponentResource {
+    public readonly instance: aws.cloudformation.Stack;
+
+    public readonly cluster: mod.Cluster;
+
+    /**
+     * The launch configuration for this auto scaling group.
+     */
+    public readonly launchConfiguration: ClusterAutoScalingLaunchConfiguration;
+
+    constructor(name: string,
+                args: ClusterAutoScalingGroupArgs,
+                opts: pulumi.ComponentResourceOptions = {}) {
+        super("aws-infra:x:ClusterAutoScalingGroup", name, args, opts);
+
+        const parentOpts = { parent: this };
+        const cluster = args.cluster;
+
+        let launchConfiguration: ClusterAutoScalingLaunchConfiguration;
+
+        // Use the autoscaling config provided, otherwise just create a default one for this cluster.
+        if (args.launchConfiguration) {
+            launchConfiguration = args.launchConfiguration;
+        }
+        else {
+            launchConfiguration = new ClusterAutoScalingLaunchConfiguration(
+                name, args.launchConfigurationArgs || { cluster }, parentOpts);
```

Comment:
another pattern.  We allow user to provide either `launchconfiguration' (the resource) or `launchconfigurationargs`.  That way, they can easily construct child objects, with a simple args object, without having to instantiate the value inline.  We may not want this as it's unclear how deep this should go, and if we shuld supply this for all  child resources.

A very common one this is good for though is "Service+TaskDefinition".  it would be somewhat annoying (IMO) when creating a Service to manually have to make the underlying TaskDefinitions.  Instead, you can just inline the TaskDefArgs into the ServiceArgs.  We should discuss how we feel about this.

Reply:
Note; my feeling is leaning toward us not doing this.  The downside though is parenting/args.  With the pattern above if you make ClusterAutoScalingGroup it can automatically pass its cluster to the ClusterAutoScalingLaunchConfiguration.  It will also parent the launch config with itself. 

If we require the user to manually instantiate this child, then they have to repetitively pass that cluster along and have to figure out a reasonable parent.  Not sure if this makes it worth it to do things this way.